### PR TITLE
Fix many GCC compiler warnings and remove unused variables

### PR DESCRIPTION
**Pull request by JohnoKing**
Most of the changes in this pull request remove unused variables, functions and labels to fix `-Wunused` compiler warnings. These are the somewhat notable changes:

- src/cmd/ksh93/bltins/print.c:
  \- Removed the unused `neg` variable (patch from ksh2020: https://github.com/att/ast/pull/725).

- src/cmd/ksh93/bltins/sleep.c:
  \- Initialized `ns` to fix three `-Wsometimes-uninitialized` warnings.

- src/cmd/ksh93/include/edit.h:
  \- Increased the size of `e_search` by one byte to fix two GCC 10 `-Wstringop-truncation` warnings.

- src/cmd/ksh93/include/shell.h:
  \- The `NOT_USED` macro caused many `-Wunused-value` warnings, so it has been replaced with ksh2020's macro: https://github.com/att/ast/commit/19d0620a (to reduce code changes, the name of `NOT_USED` hasn't been changed).

- src/cmd/ksh93/sh/expand.c:
  \- Removed an unnecessary `ap = ` since `ap` is never read between `stakseek` and `stakfreeze`.

- src/cmd/ksh93/sh/nvdisc.c, src/lib/libast/regex/regsubexec.c:
  \- Fixed some indentation to silence `-Wmisleading-indentation` warnings.

### CODE REVIEW

File: src/cmd/ksh93/bltins/typeset.c [hunk 1]
```diff
@@ -275,8 +268,8 @@ int    b_typeset(int argc,register char *argv[],Shbltin_t *context)
 					flag |= NV_HEXFLOAT;
 				}
 				else
-				        /* n=='F' Remove possible collision with NV_UNSIGN/NV_HEXFLOAT */
-				        flag &= ~NV_HEXFLOAT;
+					/* n=='F' Remove possible collision with NV_UNSIGN/NV_HEXFLOAT */
```
Comment (hyenias) [inline src/cmd/ksh93/bltins/typeset.c:271]: I did this code. Please teach me why this is wrong?

File: src/cmd/ksh93/bltins/typeset.c [hunk 2]
```diff
@@ -366,10 +359,10 @@ int    b_typeset(int argc,register char *argv[],Shbltin_t *context)
...
 				break;
 			case 'u':
-			        if(!isfloat)
-			        {
-				        tdata.wctname = e_toupper;
-				        flag |= NV_LTOU;
+				if(!isfloat)
+				{
+					tdata.wctname = e_toupper;
+					flag |= NV_LTOU;
```
Comment (hyenias) [inline src/cmd/ksh93/bltins/typeset.c:365]: I am the author of this code. Please teach me what I did wrong.

File: src/cmd/ksh93/include/shell.h [hunk 3]
```diff
@@ -37,8 +37,18 @@
 #   include	<nval.h>
 #endif /* _SH_PRIVATE */
 
+/*
+ * This is a macro that can be used to silence "unused parameter" warnings from the compiler for
+ * functions which need to accept parameters they do not use because they need to be compatible
+ * with an interface. It's similar to the Python idiom of doing `_ = param` at the top of a
+ * function in the same situation.
```
Comment (McDutchie) [inline src/cmd/ksh93/include/shell.h]: I think we could do without the Python idiom comparison in this comment.

File: src/cmd/ksh93/bltins/typeset.c (hunk 1 above)
Comment (JohnoKing) [reply src/cmd/ksh93/bltins/typeset.c:271]: You used both spaces and tabs to indent. I replaced the spaces with tabs for consistency. From policy 7 in the README:
> Your changes should match the style of the code surrounding it. Indent with **tabs**, assuming an 8-space tab width.

File: src/cmd/ksh93/bltins/typeset.c (hunk 2 above)
Comment (JohnoKing) [reply src/cmd/ksh93/bltins/typeset.c:365]: Same problem with spaces and tabs as show above.

**Review by McDutchie — CHANGES_REQUESTED** [review]
This PR removes a lot of version ID strings of the type
```c
static const char id[] = "\n@(#)$Id: ast (AT&T Research) 2012-05-28 $\0\n";
```
Please leave all those alone. They are designed for grepping the binary using [`what`(1)](https://www.freebsd.org/cgi/man.cgi?query=what&sektion=1&manpath=FreeBSD+12.2-RELEASE+and+Ports). You can do `what arch/*/bin/ksh` to see whether a particular file is linked into the binary (not all of libast is used).  After the rest of the warnings are fixed, instead of removing these ID strings I'd be inclined to add a `#pragma GCC diagnostic ignored "-Wunused-value"` directive to `src/lib/libast/include/ast.h` where, as you can see, I've already [told clang to stop complaining so much](https://github.com/ksh93/ksh/blob/83630f9d1c10330141f69e4a891f9cf0bc8dea94/src/lib/libast/include/ast.h#L23-L25) (it is a lot more eager with warnings than gcc). Alternatively, a `NOT_USED` macro invocation could be used on those version strings, but I'm not quite sure how because those are not defined in any function.

File: src/cmd/ksh93/bltins/typeset.c (hunk 1 above)
Comment (McDutchie) [reply src/cmd/ksh93/bltins/typeset.c:271]: Did those indentations actually produce gcc warnings? If not, it's hardly a big problem and I'm not sure I would have bothered, but it doesn't hurt to fix it either.

**Comment by JohnoKing** [comment]
I have reverted the removal of version ID strings and added the requested pragma to `ast.h`.

File: src/cmd/ksh93/bltins/typeset.c (hunk 1 above)
Comment (JohnoKing) [reply src/cmd/ksh93/bltins/typeset.c:271]: These indentations weren't the ones causing warnings (the indentations that caused warnings were in `nvdisc.c` and `regsubexec.c`).

**Comment by McDutchie** [comment]
On second thought I don't feel happy with the extra pragma I suggested. We risk masking future problems. We should get rid of the clang one too. Bear with me a bit, I'll fix it.

**Comment by McDutchie** [comment]
After my commit, two warnings are left when compiling with clang on macOS:
```
+ cc -D_BLD_DLL -fPIC -D_BLD_ast '-D_AST_git_commit="9501551f"' -Os -g -D_std_malloc -I. -I/usr/local/src/ksh93/ksh/src/lib/libast -
/usr/local/src/ksh93/ksh/src/lib/libast/aso/aso.c:856:10: warning: incompatible integer to pointer conversion passing 'uint64_t' (ak
                return _aso_casptr((void**)p, o, n);
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./FEATURE/aso:19:66: note: expanded from macro '_aso_casptr'
#define _aso_casptr(p,o,n)      ((void*)__sync_val_compare_and_swap(p,(uint64_t)o,(uint64_t)n))
                                                                      ^~~~~~~~~~~
/usr/local/src/ksh93/ksh/src/lib/libast/aso/aso.c:856:10: warning: incompatible integer to pointer conversion passing 'uint64_t' (ak
                return _aso_casptr((void**)p, o, n);
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
./FEATURE/aso:19:78: note: expanded from macro '_aso_casptr'
#define _aso_casptr(p,o,n)      ((void*)__sync_val_compare_and_swap(p,(uint64_t)o,(uint64_t)n))
                                                                                  ^~~~~~~~~~~
2 warnings generated.
```
I'm having a hard time figuring out what's going on there. Any clues? When I google for documentation on the `__sync_val_compare_and_swap()` atomic built-in for gcc and clang, it looks to me like it should accept any type for its parameters, but I might well be wrong.

**Comment by McDutchie** [comment]
> src/cmd/ksh93/include/edit.h:
> - Increased the size of 'e_search' by one byte to fix two
>   GCC 10 -Wstringop-truncation warnings.

Did that refer to these two `strncpy()` calls?
```
src/cmd/ksh93/edit/vi.c:2326:	strncpy(lsearch, ((char*)virtual)+1, SEARCHSIZE);
src/cmd/ksh93/edit/emacs.c:1347:		strncpy(lstring,((char*)string)+2,SEARCHSIZE);
```
The other parts of the code do seem to handle the size correctly:
```
src/cmd/ksh93/edit/vi.c:1642:			if(!vp->ed->hlist && cur_virt>=0  && cur_virt<(SEARCHSIZE-2) && cur_virt == last_virt)
src/cmd/ksh93/edit/vi.c:1644:			if(cur_virt>=0  && cur_virt<(SEARCHSIZE-2) && cur_virt == last_virt)
src/cmd/ksh93/edit/vi.c:2327:	lsearch[SEARCHSIZE-1] = 0;
src/cmd/ksh93/edit/emacs.c:1083:				if(!ep->ed->hlist && cur>0 && eol==cur && (cur<(SEARCHSIZE-2) || ep->prevdirection == -2))
src/cmd/ksh93/edit/emacs.c:1085:				if(cur>0 && eol==cur && (cur<(SEARCHSIZE-2) || ep->prevdirection == -2))
src/cmd/ksh93/edit/emacs.c:1348:		lstring[SEARCHSIZE-1] = 0;
```
Particularly, the terminating zero byte is explicitly set immediately after the two strncpy() calls. So I think it's a better fix to change the third argument to the two `strncpy()` calls to `SEARCHSIZE-1`. Can you confirm that this will silence those warnings on GCC 10?

**Comment by JohnoKing** [comment]
> Particularly, the terminating zero byte is explicitly set immediately after the two strncpy() calls. So I think it's a better fix to change the third argument to the two `strncpy()` calls to `SEARCHSIZE-1`. Can you confirm that this will silence those warnings on GCC 10?

`SEARCHSIZE-1` does silence the warnings, so this change is now used in the pull request instead.

**Pull request merged**